### ISSUE TITLE
'_RSAPublicKey' object has no attribute 'private_bytes'

### ISSUE THREAD

**Issue by vakorol** [paramiko/paramiko#748]
After upgrading to paramiko 2.0, i see the following error:

```
File "/www/deploy/lib/sshclient.py" in getPrivateKey
  146.             self.pkey.write_private_key( _fstream )
File "/usr/local/lib/python3.4/dist-packages/paramiko/rsakey.py" in write_private_key
  142.             password=password
File "/usr/local/lib/python3.4/dist-packages/paramiko/pkey.py" in _write_private_key
  347.         f.write(key.private_bytes(

Exception Type: AttributeError at /resources/2/verify/
Exception Value: '_RSAPublicKey' object has no attribute 'private_bytes'
```

This happens when i try to run method `write_private_key( _fstream )` on a `RSAKey` instance. The variable `_fstream` is an instance of a file-like stream class, child of `io.StringIO`.

This code worked fine in paramiko 1.17.

Python version: 3.4.3

**Comment by vakorol** [comment]
okay, maybe this is the same as #741

**Comment by vakorol** [comment]
This issue is really stopping us from upgrading Paramiko in our environment. Is there any chance that this will be fixed?

**Comment by bitprophet** [comment]
Yea this sounds like #741. Apologies for the long delay in getting the fix out, I intend to do a big round-up of bug milestones in the very near future!

**Comment by bitprophet** [comment]
@vakorol Please post the output of `pip list` or similar? Our tests/CI aren't reproducing this currently and the attribute in question is on a Cryptography-level object, not one of ours. Glancing over #741, I suspect it's not in fact related and am unsure why myself and others thought so at the time.

Quick trace notes:

- `paramiko.RSAKey.__init__` sets `self.key` to result of `cryptography.hazmat.backends.*.rsa.RSAPublicNumbers`
- `RSAKey.write_private_key` calls `self._write_private_key` with `key=self.key`
- `PKey._write_private_key` (note: superclass impl) takes that `key` and tries calling `private_bytes` on it

Implication being that the API of the result of that `RSAPublicNumbers` call may have fluctuated at some point.

Seems right now Paramiko claims to support Cryptography >=1.5, as of Paramiko 2.3.0; between Paramiko 2.0.0 and 2.3.0, we supported Cryptography >=1.1. I'd bet that if you went from 1.18.x to 2.x you may simply not even be updating your dependencies correctly? Definitely check there first!

**Comment by vakorol** [comment]
Thanks for the reply. I have cryptography==1.6 installed, so when installing paramiko==2.4.1, it doesn't get upgraded. Bcrypt and pynacl are installed together with the new version of paramiko correctly. Pyasn1 is installed at version 0.1.9.

When i first opened this ticket in 2016, i was trying to upgrade from 1.17 to 2.0, with the same effect.

Maybe I'm worng, but looking into [Cryptography's code](https://github.com/pyca/cryptography/blob/1.6/src/cryptography/hazmat/backends/openssl/rsa.py), it is clear that `private_bytes` is a method of `_RSAPrivateKey`, and not `_RSAPublicKey`...

**Comment by vakorol** [comment]
In any case, here is the full output of my "pip3 list":

amqp (1.4.9)
anyjson (0.3.3)
appdirs (1.4.0)
Babel (2.4.0)
backports-abc (0.5)
beautifulsoup4 (4.5.3)
billiard (3.3.0.23)
bitarray (0.8.1)
celery (3.1.23)
cffi (1.9.1)
chardet (2.2.1)
colorama (0.2.5)
cryptography (1.6)
cycler (0.10.0)
Django (1.8.19)
django-appconf (1.0.2)
django-bootstrap3 (7.0.1)
django-bulk-update (1.1.10)
django-celery-email (1.1.5)
django-classy-tags (0.8.0)
django-cms (3.4.2)
django-compressor (2.1.1)
django-formtools (2.0)
django-htmlmin (0.10.0)
django-kombu (0.9.4)
django-redis (4.8.0)
django-sekizai (0.10.0)
django-solo (1.1.2)
django-treebeard (4.1.0)
djangocms-admin-style (1.2.6.2)
djangocms-text-ckeditor (3.4.0)
djangorestframework (3.6.4)
djangorestframework-msgpack (1.0.2)
ecdsa (0.13)
flower (0.9.2)
html5lib (0.9999999)
idna (2.1)
kombu (3.0.37)
lark-parser (0.4.1)
matplotlib (1.5.1)
msgpack-python (0.4.7)
mysqlclient (1.3.7)
numpy (1.11.0)
olefile (0.44)
packaging (16.8)
paramiko (1.18.5)
Pillow (4.0.0)
pip (9.0.3)
psutil (4.1.0)
pyasn1 (0.1.9)
pycparser (2.17)
pycrypto (2.6.1)
pyminifier (2.1)
pyparsing (2.1.1)
python-dateutil (2.5.3)
pytz (2016.4)
rabbitmq (0.2.0)
rcssmin (1.0.6)
redis (2.10.5)
requests (2.2.1)
rjsmin (1.0.12)
setuptools (28.8.0)
six (1.10.0)
tornado (4.5.1)
tzlocal (1.2.2)
urllib3 (1.7.1)
uWSGI (2.0.12)
wheel (0.24.0)

**Comment by bitprophet** [comment]
> Maybe I'm worng, but looking into Cryptography's code, it is clear that private_bytes is a method of _RSAPrivateKey, and not _RSAPublicKey...

That does seem like it'd be an obvious cause, doesn't it? Not sure if this means something's mixed up. Still don't get why this isn't popping up for me, other users or the test suite though. I don't have time to dive deeper into this right now unfortunately, but a next step would be to triple check the public-vs-private aspects of the blobs in play in case there's some simple dumb mistake at work.

**Comment by vakorol** [comment]
So, i finally figured out the problem and was able to upgrade to v2.4.1.

The error I got earlier resulted from the new behavior of `paramiko.rsakey.RSAKey` constructor (compared to v1.8). In version 1.8, invoking the constructor with an instance of `paramiko.message.Message` would simply set the `e` and `n` attributes on the instance. In version 2.X, however, an instance of `rsa.RSAPublicNumbers` is created. This made the call to `write_private_key` fail ("'_RSAPublicKey' object has no attribute 'private_bytes'").

I changed my code and now it works. I assume, nobody else tried to call `RSAKey( Message(..) )`?
